modules/network_synchronizer/networked_controller.cpp

```diff
@@ -0,0 +1,1336 @@
+	scene_rewinder = p_rewinder;
+}
+
+SceneRewinder *NetworkedController::get_scene_rewinder() const {
+	return scene_rewinder;
+}
+
+bool NetworkedController::has_scene_rewinder() const {
+	return scene_rewinder;
+}
+
+void NetworkedController::_rpc_server_send_inputs(Vector<uint8_t> p_data) {
+	ERR_FAIL_COND(get_tree()->is_network_server() == false);
+
+	const Vector<int> &peers = get_active_doll_peers();
+	for (int i = 0; i < peers.size(); i += 1) {
+		// This is an active doll, Let's send the data.
+		const int peer_id = peers[i];
+		rpc_unreliable_id(peer_id, "_rpc_doll_send_inputs", p_data);
+	}
+
+	controller->receive_inputs(p_data);
+}
+
+void NetworkedController::_rpc_send_tick_additional_speed(int p_speed) {
+	ERR_FAIL_COND(is_player_controller() == false);
+
+	real_t additional_speed = (static_cast<real_t>(p_speed) / 100.0) * MAX_ADDITIONAL_TICK_SPEED;
+	additional_speed = CLAMP(additional_speed, -MAX_ADDITIONAL_TICK_SPEED, MAX_ADDITIONAL_TICK_SPEED);
+
+	PlayerController *player_controller = static_cast<PlayerController *>(controller);
+	player_controller->tick_additional_speed = additional_speed;
+}
+
+void NetworkedController::_rpc_doll_send_inputs(Vector<uint8_t> p_data) {
+	ERR_FAIL_COND_MSG(get_tree()->is_network_server() == true, "This controller is not supposed to receive this call, make sure the controllers node have the same name across all peers.");
+	ERR_FAIL_COND_MSG(is_network_master() == true, "This controller is not supposed to receive this call, make sure the controllers node have the same name across all peers.");
+
+	controller->receive_inputs(p_data);
+}
+
+void NetworkedController::_rpc_doll_notify_connection_status(bool p_open) {
+	ERR_FAIL_COND_MSG(get_tree()->is_network_server() == true, "This controller is not supposed to receive this call, make sure the controllers node have the same name across all peers.");
+	ERR_FAIL_COND_MSG(is_network_master() == true, "This controller is not supposed to receive this call, make sure the controllers node have the same name across all peers.");
+
+	if (p_open) {
+		static_cast<DollController *>(controller)->open_flow();
+	} else {
+		static_cast<DollController *>(controller)->close_flow();
+	}
+}
+
+void NetworkedController::process(real_t p_delta) {
+	if (controller) {
+		// This is called by the `sceneRewinder` that it's not aware about the
+		// controller state; so check that the controller is not null.
+		controller->physics_process(p_delta);
+	}
+}
+
+void NetworkedController::player_set_has_new_input(bool p_has) {
+	has_player_new_input = p_has;
+}
+
+bool NetworkedController::player_has_new_input() const {
+	return has_player_new_input;
+}
+
+void NetworkedController::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_READY: {
+			if (Engine::get_singleton()->is_editor_hint())
+				return;
+
+			// Unreachable.
+			CRASH_COND(get_tree() == NULL);
+
+			if (get_tree()->get_network_peer().is_null()) {
+				controller_type = CONTROLLER_TYPE_NONET;
```

Comment:
Shouldn't that constant be `CONTROLLER_TYPE_NONE`?

Reply:
`CONTROLLER_TYPE_NONE` seems that the variable is `null`. Instead it is of type No network.

I can consider rename it to: `CONTROLLER_TYPE_NONETWORK`

Reply:
Done